### DISCUSSION TITLE
User .NET 8, is there a way to not use Minimal API?

### DISCUSSION THREAD

**Discussion by dataangel** [dotnet/aspnetcore#53368] [Q&A]
I prefer the previous pattern.

**Comment by SerratedSharp** [comment]
I feel your pain.  When using a ASP.NET Core Web API project template, the Additional Information page has a "Use controllers" checkbox that should give you your desired result when checked.  Note the tooltip "...instead of minimal APIs".

![image](https://github.com/dotnet/aspnetcore/assets/97156524/fe2b5705-19ff-4685-a903-c00fea99cdf9)

![image](https://github.com/dotnet/aspnetcore/assets/97156524/c2a69bd9-18ba-4a01-8024-443c7a97603b)

**Reply by dataangel** [reply]
May I inquire about the style of this template?
//Program.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace WebSource
{
    public class Program
    {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                });
    }
}



//Startup.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Text.Encodings.Web;
using System.Text.Unicode;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authentication.Cookies;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.HttpsPolicy;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.ResponseCompression;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Origin.Web.Cache;
using Origin.Web.Http;
using Origin.Web.HttpContextExtensions;
using Origin.Web.MongoDb;
using Microsoft.AspNetCore.Mvc.NewtonsoftJson;
namespace WebSource
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            var builder = new ConfigurationBuilder().AddConfiguration(configuration)
            .AddJsonFile("appservicessettings.json", optional: true, reloadOnChange: true);
            Configuration = builder.Build();
        }
        public static string WebRootPath { get; private set; }

        public static string MapPath(string path, string basePath = null)
        {
            if (string.IsNullOrEmpty(basePath))
            {
                basePath = WebRootPath;
            }

            path = path.Replace("~/", "").TrimStart('/').Replace('/', '\\');
            return Path.Combine(basePath, path);
        }
        public IConfiguration Configuration { get; }
        public string AllowSpecificOrigins { get; private set; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            //services.Configure<GzipCompressionProviderOptions>(options => options.Level = CompressionLevel.Fastest);
            //services.AddResponseCompression(options =>
            //{
            //    options.Providers.Add<GzipCompressionProvider>();
            //});
            services.AddControllers().AddJsonOptions(options =>
            {
                options.JsonSerializerOptions.Encoder = JavaScriptEncoder.Create(UnicodeRanges.All);
            });
            services.Configure<IdentityOptions>(options =>
            {
                // Password settings
                options.Password.RequireDigit = false;
                options.Password.RequiredLength = 1;
                options.Password.RequireNonAlphanumeric = false;
                options.Password.RequireUppercase = false;
                options.Password.RequireLowercase = false;
                options.Password.RequiredUniqueChars = 1;

                // Lockout settings
                options.Lockout.DefaultLockoutTimeSpan = TimeSpan.FromMinutes(30);
                options.Lockout.MaxFailedAccessAttempts = 10;
                options.Lockout.AllowedForNewUsers = true;

                // User settings
                options.User.RequireUniqueEmail = false;
            });
            services.ConfigureApplicationCookie(options =>
            {
                options.Cookie.SameSite = SameSiteMode.Strict;
                options.AccessDeniedPath = Configuration["Cookie:AccessDeniedPath"];
                options.Cookie.Name = "Origin.Authentication.Coookie";
                options.Cookie.HttpOnly = true;
                options.ExpireTimeSpan = TimeSpan.FromDays(Convert.ToInt32(Configuration["Cookie:ExpireTimeSpan"]));
                options.LoginPath = Configuration["Cookie:LoginPath"];
                options.ReturnUrlParameter = CookieAuthenticationDefaults.ReturnUrlParameter;
                options.SlidingExpiration = true;

            });
            services.AddCors(options =>
            {
                options.AddPolicy(name: "AllowSpecificOrigins",
                                  builder =>
                                  {
                                      builder.WithOrigins(Configuration["Policy:WithOrigins"].Split(','))
                                             .AllowAnyHeader()
                                             .AllowAnyMethod().AllowCredentials(); ;
                                  });
            });
            //options.Domain = ".tnblog.net"; //设置domain让cookie共享
            services.AddMongoDbServices(x =>
            {
                x.ConnectionString = Configuration["MongoDbServices:ConnectionString"];
                x.DatabaseName = Configuration["MongoDbServices:DatabaseName"];
            });
            services.AddRedisCache(x =>
            {
                x.ConnectionString = Configuration["RedisCache:ConnectionString"];
                x.DatabaseName = Configuration["RedisCache:DatabaseName"];
            });

            services.AddSession(i =>
            {
                i.Cookie.Name = "Origin.Session";
                i.IdleTimeout = TimeSpan.FromDays(Convert.ToInt32(Configuration["Session:IdleTimeout"]));
            });

            services.AddSingleton<IHttpContextAccessor, HttpContextAccessor>();
            Context.Configure(services.BuildServiceProvider().GetRequiredService<IHttpContextAccessor>());
            //必须加入AddNewtonsoftJson才能正常格式化
            services.AddRazorPages().AddNewtonsoftJson();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            //app.UseResponseCompression();
            app.UseStaticFiles();
            app.UseRouting();
            app.UseAuthorization();
        
            WebRootPath = env.WebRootPath;
            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllerRoute(
                   name: "default",
                   pattern: "{controller=Home}/{action=Index}/{id?}");

                endpoints.MapControllerRoute(
                  name: "Area",
                  pattern: "{area:exists}/{controller=Home}/{action=Index}/{id?}");
            });
        }
    }
}
